### PR TITLE
Keep input time unit when parsing TimeValues

### PR DESCRIPTION
**Pull request by jasontedor**
This commit modifies TimeValue parsing to keep the input time unit. This
enables round-trip parsing from instances of String to instances of
TimeValue and vice-versa. With this, this commit removes support for the
unit "w" representing weeks, and also removes support for fractional
values of units (e.g., 0.5s).

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 1]
```diff
@@ -144,7 +144,7 @@ public static HighlightBuilder highlight() {
 
     private Float minScore;
 
-    private long timeoutInMillis = -1;
+    private TimeValue timeout = TimeValue.MINUS_ONE;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: I think null makes some sense here.

**Comment by nik9000** [comment]
Just to throw a wrench in this: TimeValue throws out its units when it is streamed. It is always read as nanos.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: Sadly, we use `-1` internally to mean "no timeout".

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: That makes sense when the field is a `long` but when you change it to some other type it makes less sense. I guess I'm wondering if now is the right time to change that. At least here.

File: core/src/test/java/org/elasticsearch/common/unit/TimeValueTests.java [hunk 2]
```diff
@@ -135,13 +134,15 @@ private void assertEqualityAfterSerialize(TimeValue value, int expectedSize) thr
...
 
     public void testSerialize() throws Exception {
-        assertEqualityAfterSerialize(new TimeValue(100, TimeUnit.DAYS), 8);
-        assertEqualityAfterSerialize(timeValueNanos(-1), 1);
-        assertEqualityAfterSerialize(timeValueNanos(1), 1);
-        assertEqualityAfterSerialize(timeValueSeconds(30), 6);
+        assertEqualityAfterSerialize(new TimeValue(100, TimeUnit.DAYS), 3);
+        assertEqualityAfterSerialize(timeValueNanos(-1), 2);
+        assertEqualityAfterSerialize(timeValueNanos(1), 2);
+        assertEqualityAfterSerialize(timeValueSeconds(30), 2);
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/unit/TimeValueTests.java]: Could you add a randomized example here?

File: docs/reference/migration/migrate_5_0/settings.asciidoc [hunk 3]
```diff
@@ -286,3 +286,9 @@ The setting `bootstrap.mlockall` has been renamed to
 The default setting `include_global_state` for restoring snapshots has been
 changed from `true` to `false`. It has not been changed for taking snapshots and
 still defaults to `true` in that case.
+
+==== Time value parsing
+
+The unit 'w' representing weeks is no longer supported.
+
+Fractional time values (e.g., 0.5s) are no longer supported.
```
Comment (nik9000) [inline docs/reference/migration/migrate_5_0/settings.asciidoc]: I'm not sure what'd be the right way, but it'd be nice to call out that means that things like timeouts no longer support `.5s` and should instead be `500ms` or whatever.

Comment (jasontedor) [inline docs/reference/migration/migrate_5_0/settings.asciidoc]: I pushed 3a358bf3c386db83656e32b4132f6be82a99331e.

File: core/src/test/java/org/elasticsearch/common/unit/TimeValueTests.java (hunk 2 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/common/unit/TimeValueTests.java]: I pushed 9d3b09fb4f53988a7f36ad3186d1a7094c37bb9a.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: I pushed 1e02baf900acfbcccbcd8035cbbc7cbc6e23a3c0.

**Comment by jasontedor** [comment]
> Just to throw a wrench in this: TimeValue throws out its units when it is streamed. It is always read as nanos.

I pushed 25ce3b3fe1b0974825706a9bd6c2f6008ca4db63 to modify the serialization of `TimeValue` so that this is not the case.

**Comment by nik9000** [comment]
LGTM

**Comment by jasontedor** [comment]
Thanks @nik9000!

**Comment by clintongormley** [comment]
@jasontedor out of interest, why is `w` no longer supported, given that it has a fixed duration?

**Comment by clintongormley** [comment]
Also, this page needs updating: https://www.elastic.co/guide/en/elasticsearch/reference/master/common-options.html#time-units as does this https://www.elastic.co/guide/en/elasticsearch/reference/master/search-aggregations-bucket-diversified-sampler-aggregation.html#_no_specialized_syntax_for_geo_date_fields

**Comment by jasontedor** [comment]
> out of interest, why is w no longer supported, given that it has a fixed duration?

@clintongormley This is due to the fact that we internally use the Java `TimeUnit` type to represent the time unit on a `TimeValue` instance and `TimeUnit` does not have support for weeks. Previously we converted everything internally to milliseconds and used `TimeUnit.MILLISECONDS` internally after parsing a user-input string but with this change we now parse the user-input time unit into the correct value of `TimeUnit`. To support weeks with `TimeValue` we would have to move away from the internal `TimeUnit` representation and use our own type to represent the unit. I saw the costs of this outweighing the benefit of supporting weeks. Let me know if you think otherwise?

**Comment by jasontedor** [comment]
@clintongormley Thanks; I will update the docs accordingly.

**Comment by clintongormley** [comment]
Weeks is easy to express in days, so I'm good with that

**Comment by jasontedor** [comment]
> Weeks is easy to express in days, so I'm good with that

Yes, that was in my thinking as well. Thanks @clintongormley. I'll open the PR for the docs soon.

**Comment by jasontedor** [comment]
@clintongormley I opened #19159 to update the docs.

**Pull request merged**